src/tests/test_tls_session_manager.cpp

```diff
@@ -0,0 +1,1045 @@
+            cbs.tls_current_timestamp()
+         );
+      }
+   else
+      {
+#if defined(BOTAN_HAS_TLS_13)
+      return Botan::TLS::Session(
+            {}, std::nullopt, 0, std::chrono::seconds(1024),
+            Botan::TLS::Protocol_Version::TLS_V13,
+            Botan::TLS::Ciphersuite::from_name("AES_128_GCM_SHA256")->ciphersuite_code(),
+            side,
+            {}, server_info,
+            cbs.tls_current_timestamp()
+         );
+#else
+  throw Botan_Tests::Test_Error("TLS 1.3 is not available in this build");
+#endif
+      }
+   }
+
+using namespace std::literals;
+
+std::vector<Test::Result> test_session_manager_in_memory()
+   {
+   const Botan::TLS::Session_ID default_id = random_id();
+
+   std::optional<Botan::TLS::Session_Manager_In_Memory> mgr;
+
+   Session_Manager_Callbacks cbs;
+   Session_Manager_Policy plcy;
+
+   return
+      {
+      Botan_Tests::CHECK("creation", [&](auto&)
+         {
+         mgr.emplace(Test::rng(), 5);
+         }),
+
+      Botan_Tests::CHECK("empty cache does not obtain anything", [&](auto& result)
+         {
+         result.confirm("no session found via server info", mgr->find(server_info, cbs, plcy).empty());
+
+         Botan::TLS::Session_ID mock_id = random_id();
+         auto mock_ticket = Test::rng().random_vec<Botan::TLS::Session_Ticket>(128);
+
+         result.confirm("no session found via ID", !mgr->retrieve(mock_id, cbs, plcy));
+         result.confirm("no session found via ID", !mgr->retrieve(mock_ticket, cbs, plcy));
+         }),
+
+      Botan_Tests::CHECK("clearing empty cache", [&](auto& result)
+         {
+         result.test_eq("does not delete anything", mgr->remove_all(), 0);
+         }),
+
+      Botan_Tests::CHECK("establish new session", [&](auto& result)
+         {
+         auto handle = mgr->establish(default_session(Botan::TLS::Connection_Side::Server, cbs), default_id);
+         if(result.confirm("establishment was successful", handle.has_value()))
+            {
+            result.require("session id was set", handle->id().has_value());
+            result.confirm("session ticket was empty", !handle->ticket().has_value());
+            result.test_is_eq("session id is correct", handle->id().value(), default_id);
+            }
+         }),
+
+      Botan_Tests::CHECK("obtain session from server info", [&](auto& result)
+         {
+         auto sessions = mgr->find(server_info, cbs, plcy);
+         if(result.confirm("session was found successfully", sessions.size() == 1))
+            {
+            result.test_is_eq("protocol version was echoed", sessions[0].first.version(), Botan::TLS::Protocol_Version(Botan::TLS::Version_Code::TLS_V12));
+            result.test_is_eq("ciphersuite was echoed", sessions[0].first.ciphersuite_code(), uint16_t(0x009C));
+            result.test_is_eq("ID was echoed", sessions[0].second.id().value(), default_id);
+            result.confirm("ticket was dropped", !sessions[0].second.ticket().has_value());
```

Comment:
From my understanding it never contains a ticket, so the "dropped" is a little confusing to me.

Reply:
Mhm, that's a left-over from a previous revision. At one point I thought the `Session_Handle` could host both an ID and a Ticket. I later understood that this doesn't really make sense. I'll change the label to "not a ticket".